File: 프로그래머스/lv3/131702. 고고학 최고의 발견/고고학 최고의 발견.py
from itertools import product
def rotate(clock, i,j,r):
    neighbors = [(x,y) for x,y in ((i,j),(i-1,j),(i+1,j),(i,j-1),(i,j+1)) if 0<=x<len(clock) and 0<=y<len(clock)]
    for i,j in neighbors:
        clock[i][j] = (clock[i][j] + r) % 4
def solution(clockHands):
    answer = 100000000
    for firstRotate in product(range(4), repeat=len(clockHands)):
        t=sum(firstRotate)
        clock=[[k for k in i] for i in clockHands]
        for n,r in enumerate(firstRotate): rotate(clock,0,n,r)
        for i in range(1, len(clockHands)):
            for j in range(len(clockHands)):
                if (4 - clock[i-1][j])%4:
                    t += 4 - clock[i-1][j]
                    rotate(clock,i,j,4 - clock[i-1][j])
        if not sum(clock[-1]): answer= min(answer, t)
    return answer

File: 프로그래머스/lv3/131702. 고고학 최고의 발견/고고학 최고의 발견.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> clockHands) {
    int answer = 100000000;
    for (int firstRotate = 0; firstRotate < 4; firstRotate++) {
        int t = firstRotate;
        vector<vector<int>> clock = clockHands;
        for (int n = 0; n < clockHands.size(); n++) {
            if (firstRotate & (1 << n)) {
                for (int i = 0; i < clockHands.size(); i++) {
                    clock[i][n] = (clock[i][n] + 3) % 4;
                }
            }
        }
        for (int i = 1; i < clockHands.size(); i++) {
            for (int j = 0; j < clockHands.size(); j++) {
                if (clock[i - 1][j] != 3) {
                    t += 4 - clock[i - 1][j];
                    for (int k = 0; k < clockHands.size(); k++) {
                        clock[k][j] = (clock[k][j] + 4 - clock[i - 1][j]) % 4;
                    }
                }
            }
        }
        if (clock.back().back() == 0) {
            answer = min(answer, t);
        }
    }
    return answer;
}